django/contrib/postgres/indexes.py

```diff
@@ -53,17 +53,40 @@ def get_with_params(self):
 class GinIndex(PostgresIndex):
     suffix = 'gin'
 
-    def __init__(self, *, fastupdate=None, gin_pending_list_limit=None, **kwargs):
+    def __init__(self, *, fastupdate=None, gin_pending_list_limit=None, operator_class=None, **kwargs):
         self.fastupdate = fastupdate
         self.gin_pending_list_limit = gin_pending_list_limit
+        self.operator_class = operator_class
+        paths = [field.split('.') for field in kwargs['fields']]
+        kwargs['fields'] = [path[0] for path in paths]
+        self.paths = [path[1:] for path in paths]
         super().__init__(**kwargs)
 
+    def create_sql(self, model, schema_editor, using=''):
+        statement = super().create_sql(model, schema_editor, using)
+        columns = statement.parts['columns']
+        quote_name = columns.quote_name
+        columns.quote_name = lambda x: x
+        for field, path in zip(self.fields, self.paths):
+            if not path:
+                expr = quote_name(field)
+            else:
+                expr = '({} -> {})'.format(
+                    quote_name(field),
+                    ' -> '.join("'{}'".format(key) for key in path))
+            columns.rename_column_references(
+                model._meta.db_table, field, expr)
```

Comment:
This should be handled by the fields themselves, what you've written here is what a special cased `JSONFieldGinIndex.create_sql` implementation would look like.